Sources/TuistCache/Cache/FileClient.swift

```diff
@@ -0,0 +1,147 @@
+import Foundation
+import RxSwift
+import TSCBasic
+import TuistSupport
+
+enum FileClientError: LocalizedError, FatalError {
+    case urlSessionError(Error, AbsolutePath?)
+    case serverSideError(HTTPURLResponse, AbsolutePath?)
+    case invalidResponse(URLRequest, AbsolutePath?)
+    case noLocalURL(HTTPURLResponse)
+
+    // MARK: - FatalError
+
+    public var description: String {
+        var output: String
+
+        switch self {
+        case let .urlSessionError(error, path):
+            output = "Received a session error"
+            if let path = path {
+                output.append(" while uploading file at path \(path.pathString)")
+            }
+            if let error = error as? LocalizedError {
+                output.append("\nError: \(error.localizedDescription)")
+            }
+        case let .invalidResponse(urlRequest, path):
+            output = "Received unexpected response from the network with url request \(urlRequest)"
+            if let path = path {
+                output.append(" while uploading file at path \(path.pathString)")
+            }
+        case let .serverSideError(response, path):
+            output = "Got error code: \(response.statusCode) returned by the server"
+            if let path = path {
+                output.append(" when uploading file at path \(path.pathString)")
+            }
+            output.append("HTTPURLResponse: \(response.description)")
+        case let .noLocalURL(response):
+            output = "Could not locate file on disk the downloaded file. HTTPURLResponse: \(response.description)"
```

Comment:
`HTTPURLResponse` is an implementation detail not relevant to the user. If the issue is that the file cannot be found locally, we can say something like: 

```
Couldn't upload the file at path xxx because it doesn't exist locally. 
```

Reply:
Corrected.